stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -249,15 +250,23 @@ struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
                              std::forward<Vec>(vmapped),
                              std::forward<Args>(args)...);
 
-    if (auto_partitioning) {
-      tbb::parallel_reduce(
-          tbb::blocked_range<std::size_t>(0, num_terms, grainsize), worker);
-    } else {
-      tbb::simple_partitioner partitioner;
-      tbb::parallel_deterministic_reduce(
-          tbb::blocked_range<std::size_t>(0, num_terms, grainsize), worker,
-          partitioner);
-    }
+    // we must use task isolation as described here:
+    // https://software.intel.com/content/www/us/en/develop/documentation/tbb-documentation/top/intel-threading-building-blocks-developer-guide/task-isolation.html
+    // this is to ensure that the thread local AD tape ressource is
+    // not being modified from a different task which may happen
+    // whenever this function is being used itself in a parallel
+    // context (like running multiple chains for Stan)
+    tbb::this_task_arena::isolate([&] {
+      if (auto_partitioning) {
+        tbb::parallel_reduce(
+            tbb::blocked_range<std::size_t>(0, num_terms, grainsize), worker);
+      } else {
+        tbb::simple_partitioner partitioner;
+        tbb::parallel_deterministic_reduce(
+            tbb::blocked_range<std::size_t>(0, num_terms, grainsize), worker,
+            partitioner);
+      }
+    });
```

Comment:
[optional] I'd prefer explicit captures for this and the function in `map_rect` but it's nbd